src/lib/network_pool/indexed_pool.ml

```diff
@@ -475,6 +515,29 @@ let revalidate :
                     ~data:(keep_queue', currency_reserved'') }
             , Sequence.append dropped_acc to_drop ) )
 
+let remove_expired t :
+    Transaction_hash.User_command_with_valid_signature.t Sequence.t * t =
+  Map.fold t.all_by_sender ~init:(Sequence.empty, t)
```

Comment:
This appears to walk every transaction in the pool. In the protocol prio meeting, we briefly discussed adding a
`Transaction_hash.User_command_with_valid_signature.Set.t Global_slot.Map.t` to the pool state and using `Map.split` to efficiently identify any old transactions. Can you rework this to do that instead?

Reply:
Deleting expired transactions will cause any dependent transactions to also be deleted and these might not be expired but agree that it'll be better on an average with the extra map . Then there is an overhead of maintaining this new map for every other transaction pool update. So walking every transaction in the pool once every new best tip seemed like an ok trade-off to make?

Reply:
> Deleting expired transactions will cause any dependent transactions to also be deleted and these might not be expired

This is a unfortunate, and a potential DOS vector, but as soon as we don't have a valid transaction with an earlier nonce they have no value and we should be purging them anyway.

> Then there is an overhead of maintaining this new map for every other transaction pool update. So walking every transaction in the pool once every new best tip seemed like an ok trade-off to make?

Adding a new txn to the map is worst-case `O(log(n))`, as is identifying the map partition to be deleted. Doing an `O(n)` check with any regularity is exponentially worse, and I would prefer to avoid it. It will usually be even more efficient with a map, since we can avoid putting any untimed transactions in at all, so txns in the pool will be much more plentiful than in the map.

Reply:
Well, there are these operations that cause deletion of dependent transactions and would take an additional O(m log n) where m is the # of transactions to be deleted. So, wasn't exactly sure if it was better to have another map to maintain
1. `remove_lowest_fee` that'll delete a transaction and its dependants in the case of overflow. This is called both when adding a gossiped transaction and handling best tip transactions. 
2. In `add_from_gossip_exn` dropping commands when replacing a transaction. But this probably doesn't happen too often?
3. `drop_until_sufficient_balance` that deletes transactions when handling a committed transaction.

Reply:
I was disregarding the `m` because it's useful work rather than overhead: we're clearing now-useless transactions and never having to handle them again, and without doing any additional evaluations.

Removing them from the extra map should be as simple as adding `Map.update txn.valid_until ~f:(Set.remove txn)` for each of those functions.